### PR TITLE
trans: Internalize symbols without relying on LLVM

### PR DESCRIPTION
**Pull request by michaelwoerister**
This PR makes the compiler use the information gather by the trans collector in order to determine which symbols/trans-items can be made internal. This has the advantages:
+ of being LLVM independent, 
+ of also working in incremental mode, and
+ of allowing to not keep all LLVM modules in memory at the same time.

This is in preparation for fixing issue #39280.

cc @rust-lang/compiler

### CODE REVIEW

File: src/librustc_trans/back/symbol_export.rs [hunk 1]
```diff
@@ -28,59 +27,87 @@ pub enum SymbolExportLevel {
 }
 
 /// The set of symbols exported from each crate in the crate graph.
+#[derive(Clone, Debug)]
```
Comment (Mark-Simulacrum) [inline src/librustc_trans/back/symbol_export.rs]: Does this need to be `Clone`? It looks like it's a fairly large structure that ideally we wouldn't clone. It seems like it's in an Arc most of the time anyway so this feels unnecessary.

File: src/librustc_trans/collector.rs [hunk 2]
```diff
@@ -231,33 +237,61 @@ impl<'tcx> InliningMap<'tcx> {
...
+                          source: TransItem<'tcx>,
+                          targets: I)
+        where I: Iterator<Item=(TransItem<'tcx>, bool)>
     {
         assert!(!self.index.contains_key(&source));
 
         let start_index = self.targets.len();
-        self.targets.extend(targets);
+        let (targets_size_hint, targets_size_hint_max) = targets.size_hint();
+        debug_assert_eq!(targets_size_hint_max, Some(targets_size_hint));
```
Comment (Mark-Simulacrum) [inline src/librustc_trans/collector.rs]: It looks like what we really want here is an `ExactSizeIterator`?

File: src/librustc_trans/partitioning.rs [hunk 3]
```diff
@@ -433,6 +538,73 @@ fn place_inlined_translation_items<'tcx>(initial_partitioning: PreInliningPartit
...
+    let trans_item_placements = &partitioning.trans_item_placements;
+
+    // For each internalization candidates in each codegen unit, check if it is
+    // accessed from outside its defining codegen unit.
+    for cgu in &mut partitioning.codegen_units {
+        let home_cgu = TransItemPlacement::SingleCgu {
+            cgu_name: cgu.name.clone()
+        };
+
+        'item:
```
Comment (Mark-Simulacrum) [inline src/librustc_trans/partitioning.rs]: This appears to no longer be necessary.

File: src/librustc_trans/partitioning.rs [hunk 4]
```diff
@@ -433,6 +538,73 @@ fn place_inlined_translation_items<'tcx>(initial_partitioning: PreInliningPartit
...
+            }
+            debug_assert_eq!(trans_item_placements[accessee], home_cgu);
+
+            if let Some(accessors) = accessor_map.get(accessee) {
+                if accessors.iter()
+                            .filter_map(|accessor| {
+                                // Some accessors might not have been
+                                // instantiated. We can safely ignore those.
+                                trans_item_placements.get(accessor)
+                                                     .cloned()
```
Comment (Mark-Simulacrum) [inline src/librustc_trans/partitioning.rs]: Why do we need to clone the option? It looks like that shouldn't be necessary -- we only use it for `PartialEq`?

File: src/librustc_trans/collector.rs [hunk 5]
```diff
@@ -850,8 +910,14 @@ impl<'b, 'a, 'v> ItemLikeVisitor<'v> for RootCollector<'b, 'a, 'v> {
                 // const items only generate translation items if they are
                 // actually used somewhere. Just declaring them is insufficient.
             }
-            hir::ItemFn(.., ref generics, _) => {
-                if !generics.is_type_parameterized() {
+            hir::ItemFn(_, _, constness, _, ref generics, _) => {
+                let is_const = match constness {
```
Comment (eddyb) [inline src/librustc_trans/collector.rs]: I recommend `tcx.is_const_fn(def_id)` - similarly for `hir::Generics`, `ty::Generics` should be preferred, although that one is preexisting.

File: src/librustc_trans/collector.rs [hunk 6]
```diff
@@ -889,7 +956,14 @@ impl<'b, 'a, 'v> ItemLikeVisitor<'v> for RootCollector<'b, 'a, 'v> {
                     }
                 };
 
-                if !generics.is_type_parameterized() && !is_impl_generic {
+                let is_const = match constness {
```
Comment (eddyb) [inline src/librustc_trans/collector.rs]: Same here.

**Comment by eddyb** [comment]
r=me modulo nits

File: src/librustc_trans/back/symbol_export.rs (hunk 1 above)
Comment (michaelwoerister) [inline src/librustc_trans/back/symbol_export.rs]: That is indeed not necessary anymore. Will remove.

File: src/librustc_trans/collector.rs (hunk 2 above)
Comment (michaelwoerister) [inline src/librustc_trans/collector.rs]: I'll look into it.

File: src/librustc_trans/partitioning.rs (hunk 3 above)
Comment (michaelwoerister) [inline src/librustc_trans/partitioning.rs]: I don't understand. It's used in the nested `continue` below.

File: src/librustc_trans/partitioning.rs (hunk 4 above)
Comment (michaelwoerister) [inline src/librustc_trans/partitioning.rs]: `cloned()` (as opposed to `clone()`) turns the `Option<&T>` into an `Option<T>`, which is what we need for `filter_map`.

File: src/librustc_trans/collector.rs (hunk 5 above)
Comment (michaelwoerister) [inline src/librustc_trans/collector.rs]: Does `is_const_fn` do something differently than the code below? I'm not against changing, but the `constness` and `generics` from the HIR item are right there without the need for a hashtable lookup.

File: src/librustc_trans/partitioning.rs (hunk 4 above)
Comment (michaelwoerister) [inline src/librustc_trans/partitioning.rs]: I see what you meant now. Yes, cloning seems unnecessary.

File: src/librustc_trans/partitioning.rs (hunk 3 above)
Comment (Mark-Simulacrum) [inline src/librustc_trans/partitioning.rs]: I'm probably just missing something, but I don't see a nested loop above it?

Comment (michaelwoerister) [inline src/librustc_trans/partitioning.rs]: Oh right, the only place those nested loops still exist is in my head `:)` Sorry for the confusion. I'll remove the label.

**Comment by michaelwoerister** [comment]
@bors r=eddyb

All nits addressed, I think.

**Comment by Mark-Simulacrum** [comment]
```
00:16:51] rustc: /checkout/src/llvm/include/llvm/IR/GlobalValue.h:227: void llvm::GlobalValue::setVisibility(llvm::GlobalValue::VisibilityTypes): Assertion `(!hasLocalLinkage() || V == DefaultVisibility) && "local linkage requires default visibility"' failed.
[00:16:51] (B[m[31m[1merror:(B[m Could not compile `core`.
[00:16:51] (B[m
[00:16:51] Caused by:
[00:16:51] (B[m(B[m  process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustc --crate-name core /checkout/src/libcore/lib.rs --color always --error-format json --crate-type lib --emit=dep-info,link -C opt-level=2 -C metadata=a6117b7e5e0db06a -C extra-filename=-a6117b7e5e0db06a --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps --target x86_64-unknown-linux-gnu -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/release/deps -Zincremental=/tmp/rust-incr-cache` (signal: 6, SIGABRT: process abort signal)
[00:16:51] --- stderr
[00:16:51] rustc: /checkout/src/llvm/include/llvm/IR/GlobalValue.h:227: void llvm::GlobalValue::setVisibility(llvm::GlobalValue::VisibilityTypes): Assertion `(!hasLocalLinkage() || V == DefaultVisibility) && "local linkage requires default visibility"' failed.
[00:16:51] 
[00:16:51] (B[mthread 'main' panicked at 'command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "-j" "4" "--target" "x86_64-unknown-linux-gnu" "--release" "--locked" "--color" "always" "--features" "panic-unwind jemalloc backtrace" "--manifest-path" "/checkout/src/libstd/Cargo.toml" "--message-format" "json"
[00:16:51] expected success, got: exit code: 101', /checkout/src/bootstrap/compile.rs:586
[00:16:51] note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

**Comment by michaelwoerister** [comment]
@bors r=eddyb

That assertion should be fixed, hopefully.

**Comment by kennytm** [comment]
Spurious. Could this be related to https://appveyor.statuspage.io/incidents/m2vdvw39kdk8?

```
  [35] SSL connect error (schannel: next InitializeSecurityContext failed: Unknown error (0x80092013) - The revocation function was unable to check revocation because the revocation server was offline.)
```

**Comment by michaelwoerister** [comment]
@bors retry

(No harm in retrying while the queue is empty, I guess)

**Comment by kennytm** [comment]
`x86_64-pc-windows-gnu` failed to link stage1-std ("undefined reference to `std::thread::local::os::destroy_value::hXXXXXX`"). Seems legit.

```
error: linking with `gcc` failed: exit code: 1
  |
  = note: "gcc" "-Wl,--enable-long-section-names" "-fno-use-linker-plugin" "-Wl,--nxcompat" "-nostdlib" "-m64" "C:\\projects\\rust\\build\\x86_64-pc-windows-gnu\\stage1\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\dllcrt2.o" "C:\\projects\\rust\\build\\x86_64-pc-windows-gnu\\stage1\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\rsbegin.o" "-L" "C:\\projects\\rust\\build\\x86_64-pc-windows-gnu\\stage1\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib" "C:\\projects\\rust\\build\\x86_64-pc-windows-gnu\\stage1-std\\x86_64-pc-windows-gnu\\release\\deps\\std-ac667c8bbd86e364.0.o" "-o" "C:\\projects\\rust\\build\\x86_64-pc-windows-gnu\\stage1-std\\x86_64-pc-windows-gnu\\release\\deps\\std-ac667c8bbd86e364.dll" "C:\\projects\\rust\\build\\x86_64-pc-windows-gnu\\stage1-std\\x86_64-pc-windows-gnu\\release\\deps\\std-ac667c8bbd86e364.crate.metadata.o" "C:\\projects\\rust\\build\\x86_64-pc-windows-gnu\\stage1-std\\x86_64-pc-windows-gnu\\release\\deps\\std-ac667c8bbd86e364.crate.allocator.o" "-nodefaultlibs" "-L" "C:\\projects\\rust\\build\\x86_64-pc-windows-gnu\\stage1-std\\x86_64-pc-windows-gnu\\release\\deps" "-L" "C:\\projects\\rust\\build\\x86_64-pc-windows-gnu\\stage1-std\\release\\deps" "-L" "C:\\projects\\rust\\build\\x86_64-pc-windows-gnu\\native\\libbacktrace\\.libs" "-L" "C:\\projects\\rust\\build\\x86_64-pc-windows-gnu\\native\\jemalloc\\lib" "-L" "C:\\projects\\rust\\build\\x86_64-pc-windows-gnu\\stage1-std\\x86_64-pc-windows-gnu\\release\\build\\compiler_builtins-3f52fffdb7374465\\out" "-L" "C:\\projects\\rust\\build\\x86_64-pc-windows-gnu\\stage1\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib" "-Wl,-Bstatic" "-Wl,--whole-archive" "-l" "backtrace" "-Wl,--no-whole-archive" "-Wl,-Bdynamic" "-l" "advapi32" "-l" "ws2_32" "-l" "userenv" "-l" "shell32" "-Wl,-Bstatic" "-Wl,--whole-archive" "C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\rustc.HACeg3fOyh3C\\libpanic_unwind-d7ba3ed13adbba78.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\rustc.HACeg3fOyh3C\\libunwind-4a5bfaad61fe8668.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\rustc.HACeg3fOyh3C\\liblibc-f12a6f077c568615.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\rustc.HACeg3fOyh3C\\librand-fc4f55c30253478d.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\rustc.HACeg3fOyh3C\\liballoc_system-fa945e0e44cf2c85.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\rustc.HACeg3fOyh3C\\liballoc-7e25d9aea6733470.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\rustc.HACeg3fOyh3C\\libstd_unicode-b2da6d85e26577df.rlib" "-Wl,--no-whole-archive" "-Wl,--whole-archive" "C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\rustc.HACeg3fOyh3C\\libcore-133c3145709a073d.rlib" "-Wl,--no-whole-archive" "C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\rustc.HACeg3fOyh3C\\libcompiler_builtins-b49c9299458b621e.rlib" "-l" "gcc_eh" "-l" "pthread" "-shared" "-Wl,-Bdynamic" "-lmingwex" "-lmingw32" "-lgcc" "-lmsvcrt" "-luser32" "-lkernel32" "C:\\projects\\rust\\build\\x86_64-pc-windows-gnu\\stage1\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\rsend.o"
  = note: C:\projects\rust\build\x86_64-pc-windows-gnu\stage1-std\x86_64-pc-windows-gnu\release\deps\std-ac667c8bbd86e364.0.o:(.data+0xd8): undefined reference to `std::thread::local::os::destroy_value::ha44314076d36f203'
          C:\projects\rust\build\x86_64-pc-windows-gnu\stage1-std\x86_64-pc-windows-gnu\release\deps\std-ac667c8bbd86e364.0.o:(.data+0xe8): undefined reference to `std::thread::local::os::destroy_value::hc6b9798d5a01e96f'
          C:\projects\rust\build\x86_64-pc-windows-gnu\stage1-std\x86_64-pc-windows-gnu\release\deps\std-ac667c8bbd86e364.0.o:(.data+0xf8): undefined reference to `std::thread::local::os::destroy_value::hfc4011e45796e013'
          C:\projects\rust\build\x86_64-pc-windows-gnu\stage1-std\x86_64-pc-windows-gnu\release\deps\std-ac667c8bbd86e364.0.o:(.data+0x108): undefined reference to `std::thread::local::os::destroy_value::hc6b9798d5a01e96f'
          C:\projects\rust\build\x86_64-pc-windows-gnu\stage1-std\x86_64-pc-windows-gnu\release\deps\std-ac667c8bbd86e364.0.o:(.data+0x118): undefined reference to `std::thread::local::os::destroy_value::h96e432fe261f7344'
          C:\projects\rust\build\x86_64-pc-windows-gnu\stage1-std\x86_64-pc-windows-gnu\release\deps\std-ac667c8bbd86e364.0.o:(.data+0x128): undefined reference to `std::thread::local::os::destroy_value::h10f11fb8acf74f1c'
          collect2.exe: error: ld returned 1 exit status
          
error: aborting due to previous error
error: Could not compile `std`.
```

**Comment by michaelwoerister** [comment]
Thanks, @kennytm!

**Comment by aidanhs** [comment]
```
[00:40:46] ---- [compile-fail] compile-fail/E0534.rs stdout ----
[00:40:46] 	
[00:40:46] error: compile-fail test compiled successfully!
[00:40:46] status: exit code: 0
[00:40:46] command: /checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc /checkout/src/test/compile-fail/E0534.rs -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/compile-fail --target=x86_64-unknown-linux-gnu --error-format json -L /checkout/obj/build/x86_64-unknown-linux-gnu/test/compile-fail/E0534.stage2-x86_64-unknown-linux-gnu.compile-fail.libaux -C prefer-dynamic -o /checkout/obj/build/x86_64-unknown-linux-gnu/test/compile-fail/E0534.stage2-x86_64-unknown-linux-gnu -Crpath -O -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers
[00:40:46] stdout:
```
(and a few more)

**Comment by michaelwoerister** [comment]
@bors r=eddyb

@eddyb I pushed one more commit to fix the linker issue on Windows. It should be rather uncontroversial but feel free to take another look.

**Pull request merged**